From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Subject: Disable general-with-expect; which tends to fail on ppc64el and arm64.

---
 src/tests/Makefile.am |    1 -
 1 file changed, 1 deletion(-)

Index: b/src/tests/Makefile.am
===================================================================
--- a/src/tests/Makefile.am
+++ b/src/tests/Makefile.am
@@ -157,7 +157,6 @@ TESTS = \
 	test-dcb \
 	test-resolvconf-capture \
 	test-general \
-	test-general-with-expect \
 	test-systemd \
 	test-wired-defname \
 	test-utils
