From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Date: Wed, 15 Jun 2016 16:07:27 +0300
Subject: dns-manager: don't merge split-DNS search domains

Signed-off-by: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/network-manager/+bug/1592721
---
 src/dns/nm-dns-manager.c | 36 ++++++++++++++++++++----------------
 1 file changed, 20 insertions(+), 16 deletions(-)

Index: b/src/dns/nm-dns-manager.c
===================================================================
--- a/src/dns/nm-dns-manager.c
+++ b/src/dns/nm-dns-manager.c
@@ -295,7 +295,7 @@
 }
 
 static void
-merge_one_ip4_config (NMResolvConfData *rc, NMIP4Config *src)
+merge_one_ip4_config (NMResolvConfData *rc, NMIP4Config *src, gboolean never_default)
 {
 	guint32 num, num_domains, num_searches, i;
 
@@ -308,13 +308,14 @@
 	num_domains = nm_ip4_config_get_num_domains (src);
 	num_searches = nm_ip4_config_get_num_searches (src);
 
-	for (i = 0; i < num_searches; i++) {
-		const char *search;
-
-		search = nm_ip4_config_get_search (src, i);
-		if (!domain_is_valid (search, FALSE))
-			continue;
-		add_string_item (rc->searches, search);
+	if (!never_default) {
+		for (i = 0; i < num_searches; i++) {
+			const char *search;
+			search = nm_ip4_config_get_search (src, i);
+			if (!domain_is_valid (search, FALSE))
+				continue;
+			add_string_item (rc->searches, search);
+		}
 	}
 
 	if (num_domains > 1 || !num_searches) {
@@ -351,7 +352,7 @@
 }
 
 static void
-merge_one_ip6_config (NMResolvConfData *rc, NMIP6Config *src, const char *iface)
+merge_one_ip6_config (NMResolvConfData *rc, NMIP6Config *src, const char *iface, gboolean never_default)
 {
 	guint32 num, num_domains, num_searches, i;
 
@@ -378,13 +379,14 @@
 	num_domains = nm_ip6_config_get_num_domains (src);
 	num_searches = nm_ip6_config_get_num_searches (src);
 
-	for (i = 0; i < num_searches; i++) {
-		const char *search;
-
-		search = nm_ip6_config_get_search (src, i);
-		if (!domain_is_valid (search, FALSE))
-			continue;
-		add_string_item (rc->searches, search);
+	if (!never_default) {
+		for (i = 0; i < num_searches; i++) {
+			const char *search;
+			search = nm_ip6_config_get_search (src, i);
+			if (!domain_is_valid (search, FALSE))
+				continue;
+			add_string_item (rc->searches, search);
+		}
 	}
 
 	if (num_domains > 1 || !num_searches) {
@@ -412,9 +414,9 @@
                           NMDnsIPConfigData *data)
 {
 	if (NM_IS_IP4_CONFIG (data->config))
-		merge_one_ip4_config (rc, (NMIP4Config *) data->config);
+		merge_one_ip4_config (rc, (NMIP4Config *) data->config, nm_ip4_config_get_never_default (data->config));
 	else if (NM_IS_IP6_CONFIG (data->config))
-		merge_one_ip6_config (rc, (NMIP6Config *) data->config, data->iface);
+		merge_one_ip6_config (rc, (NMIP6Config *) data->config, data->iface, nm_ip4_config_get_never_default (data->config));
 	else
 		g_return_if_reached ();
 }
