From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Subject: Ignore rmnet_usbX devices.
---
 src/nm-manager.c |    5 +++++
 1 file changed, 5 insertions(+)

Index: network-manager-0.9.10.0/src/nm-manager.c
===================================================================
--- network-manager-0.9.10.0.orig/src/nm-manager.c
+++ network-manager-0.9.10.0/src/nm-manager.c
@@ -2054,6 +2054,12 @@ platform_link_added (NMManager *self,
 	if (!strncmp (plink->name, "bnep", STRLEN ("bnep")))
 		return;
 
+	/* Ignore rmnet_usb devices, which will be handled by their modem parent
+	 */
+	if (!strncmp (plink->name, "rmnet_usb", STRLEN ("rmnet_usb")) ||
+		!strncmp (plink->name, "ccmni", STRLEN ("ccmni")))
+		return;
+
 	if (device == NULL) {
 		int parent_ifindex = -1;
 		NMDevice *parent;
