From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Subject: Prioritize Wifi over 3G, if only because it's cheaper.

---
 src/devices/nm-device.c |   16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

Index: b/src/devices/nm-device.c
===================================================================
--- a/src/devices/nm-device.c
+++ b/src/devices/nm-device.c
@@ -717,6 +717,10 @@ nm_device_get_priority (NMDevice *self)
 	 * metrics (except for IPv6, where 0 means 1024).
 	 */
 
+	/* Ubuntu change:
+	 * Adjust priorities to de-prioritize bluetooth, wimax and modem
+	 * to lower; BT is just slow, WiMAX and "MODEM" because they cost
+	 * money. */
 	switch (nm_device_get_device_type (self)) {
 	/* 50 is reserved for VPN (NM_VPN_ROUTE_METRIC_DEFAULT) */
 	case NM_DEVICE_TYPE_ETHERNET:
@@ -725,8 +729,6 @@ nm_device_get_priority (NMDevice *self)
 		return 150;
 	case NM_DEVICE_TYPE_ADSL:
 		return 200;
-	case NM_DEVICE_TYPE_WIMAX:
-		return 250;
 	case NM_DEVICE_TYPE_BOND:
 		return 300;
 	case NM_DEVICE_TYPE_TEAM:
@@ -735,14 +737,16 @@ nm_device_get_priority (NMDevice *self)
 		return 400;
 	case NM_DEVICE_TYPE_BRIDGE:
 		return 425;
-	case NM_DEVICE_TYPE_MODEM:
-		return 450;
-	case NM_DEVICE_TYPE_BT:
-		return 550;
 	case NM_DEVICE_TYPE_WIFI:
 		return 600;
 	case NM_DEVICE_TYPE_OLPC_MESH:
 		return 650;
+	case NM_DEVICE_TYPE_WIMAX:
+		return 700;
+	case NM_DEVICE_TYPE_MODEM:
+		return 750;
+	case NM_DEVICE_TYPE_BT:
+		return 800;
 	case NM_DEVICE_TYPE_GENERIC:
 		return 950;
 	case NM_DEVICE_TYPE_UNKNOWN:
